# Econ 441A Assignment 4
## Juntao Li
## Student ID: 906092273

### 1/a)

In [1]:
import wooldridge as woo
import numpy as np
import statsmodels.formula.api as smf
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import scipy.stats as stats
mydata=woo.data('DISCRIM')

In [3]:
reg1=smf.ols('np.log(psoda)~prpblck+np.log(income)+prppov',data=mydata)
results1=reg1.fit()
results1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(psoda)   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.080
Method:                 Least Squares   F-statistic:                     12.60
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           6.92e-08
Time:                        18:09:38   Log-Likelihood:                 439.04
No. Observations:                 401   AIC:                            -870.1
Df Residuals:                     397   BIC:                            -854.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -1.4633      0.294     -4.982      0.000      -2.041      -0.886
prpblck            0.0728      0.031      2.373      0.018       0.013       0.133
np.log(income)     0.1370      0.027      5.119      0.000       0.084       0.190
prppov             0.3804      0.133      2.864      0.004       0.119       0.641
==============================================================================
Omnibus:                       12.002   Durbin-Watson:                   1.727
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               24.056
Skew:                          -0.014   Prob(JB):                     5.97e-06
Kurtosis:                       4.200   Cond. No.                         834.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The p-value for beta_1 is 0.018, which is lower than 0.05 but greater than 0.01, so it means beta_1 is statistically different from zero at 5% level against a two-sided alternative but is not statistically different from zero at 1% level against a two-sided alternative.

### 1/b)

In [4]:
print("Correlation is",np.log(mydata.income).corr(mydata.prppov))

Correlation is -0.8384669733579033


In [5]:
import patsy as pt
import statsmodels.stats.outliers_influence as smo

y,X = pt.dmatrices('np.log(psoda)~prpblck+np.log(income)+prppov', data = mydata,
                   return_type = 'dataframe')

X.head()

Intercept   prpblck  np.log(income)    prppov
0        1.0  0.171154       10.704008  0.036579
1        1.0  0.171154       10.704008  0.036579
2        1.0  0.047360       10.625319  0.087907
3        1.0  0.052839       10.827072  0.059123
4        1.0  0.034480       11.188400  0.025415

In [6]:
k = X.shape[1]

VIF = np.empty(k)

for i in range(k):
    
    VIF[i] = smo.variance_inflation_factor(X.values, i)

print('VIF:', VIF)

VIF: [5.22495302e+03 1.92215952e+00 3.51872083e+00 4.91521981e+00]


In [7]:
results1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(psoda)   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.080
Method:                 Least Squares   F-statistic:                     12.60
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           6.92e-08
Time:                        18:09:38   Log-Likelihood:                 439.04
No. Observations:                 401   AIC:                            -870.1
Df Residuals:                     397   BIC:                            -854.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -1.4633      0.294     -4.982      0.000      -2.041      -0.886
prpblck            0.0728      0.031      2.373      0.018       0.013       0.133
np.log(income)     0.1370      0.027      5.119      0.000       0.084       0.190
prppov             0.3804      0.133      2.864      0.004       0.119       0.641
==============================================================================
Omnibus:                       12.002   Durbin-Watson:                   1.727
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               24.056
Skew:                          -0.014   Prob(JB):                     5.97e-06
Kurtosis:                       4.200   Cond. No.                         834.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

According to the summary table above, the p-values for prpblck, log(income), prppov are 0.018,0.000,0.004,respectively, where 0.01<0.018<0.05, 0.000<0.01, 0.004<0.01. So log(income) and prppov are statistically significant under any circumstances and prpblck is not statistically significant only at 1% level.

### 1/c)

In [8]:
reg2=smf.ols('np.log(psoda)~prpblck+np.log(income)+prppov+np.log(hseval)',data=mydata)
results2=reg2.fit()
results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(psoda)   R-squared:                       0.184
Model:                            OLS   Adj. R-squared:                  0.176
Method:                 Least Squares   F-statistic:                     22.31
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           1.24e-16
Time:                        18:09:38   Log-Likelihood:                 461.55
No. Observations:                 401   AIC:                            -913.1
Df Residuals:                     396   BIC:                            -893.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept         -0.8415      0.292     -2.878      0.004      -1.416      -0.267
prpblck            0.0976      0.029      3.334      0.001       0.040       0.155
np.log(income)    -0.0530      0.038     -1.412      0.159      -0.127       0.021
prppov             0.0521      0.134      0.388      0.699      -0.212       0.317
np.log(hseval)     0.1213      0.018      6.860      0.000       0.087       0.156
==============================================================================
Omnibus:                       16.452   Durbin-Watson:                   1.973
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               40.077
Skew:                          -0.016   Prob(JB):                     1.98e-09
Kurtosis:                       4.548   Cond. No.                     1.31e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.31e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

The coefficient is 0.1213, which means 1% increase in the median housing value will increase the price of medium soda by 12.13%. And the p-value is 0.000, which is less than 0.005, it means that we need to reject H0 and the coefficient of log(hseval) is statistically significant.

### 1/d)

In [9]:
hypotheses = ["np.log(income)=0","prppov=0"]
results2.f_test(hypotheses)

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=3.522687204208135, p=0.030448537491840212, df_denom=396, df_num=2>

In the new regression model above, the p-value for log(income) is 0.159 and p-value for prppov is 0.699. Since 0.159>0.05, 0.699>0.05, log(income) and prppov are not statistically significant as individual variables.   
Since the p-value for the F-Test is 0.0304, which is less than 0.05, we reject the null that both coefficients of log(income) and prppov are equal to zero. So these two variables are jointly significant.

### 1/e)

I will choose the model from part c) since it has a greater R-squared value than the model in part a). Even though log(income) and prppov are not significant invidually, they are jointly significant, so it should be fine to keep these two variables in the model.

### 2/a)

In [10]:
mydata1=woo.data('vote1')

In [11]:
reg3=smf.ols("voteA~prtystrA+np.log(expendA)+democA+np.log(expendB)",data=mydata1)
results3=reg3.fit()
Residual=results3.resid
Residual
results3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.796
Method:                 Least Squares   F-statistic:                     169.2
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           8.09e-58
Time:                        18:09:38   Log-Likelihood:                -593.20
No. Observations:                 173   AIC:                             1196.
Df Residuals:                     168   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          37.6614      4.736      7.952      0.000      28.312      47.011
prtystrA            0.2519      0.071      3.534      0.001       0.111       0.393
np.log(expendA)     5.7793      0.392     14.750      0.000       5.006       6.553
democA              3.7929      1.407      2.697      0.008       1.016       6.570
np.log(expendB)    -6.2378      0.397    -15.694      0.000      -7.022      -5.453
==============================================================================
Omnibus:                        6.304   Durbin-Watson:                   1.525
Prob(Omnibus):                  0.043   Jarque-Bera (JB):                6.030
Skew:                           0.448   Prob(JB):                       0.0491
Kurtosis:                       3.182   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [12]:
reg4=smf.ols("Residual~prtystrA+np.log(expendA)+democA+np.log(expendB)",data=mydata1)
results4=reg4.fit()
results4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               Residual   R-squared:                      -0.000
Model:                            OLS   Adj. R-squared:                 -0.024
Method:                 Least Squares   F-statistic:                -7.929e-15
Date:                Mon, 14 Nov 2022   Prob (F-statistic):               1.00
Time:                        18:09:38   Log-Likelihood:                -593.20
No. Observations:                 173   AIC:                             1196.
Df Residuals:                     168   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept       -1.821e-14      4.736  -3.84e-15      1.000      -9.350       9.350
prtystrA         7.633e-16      0.071   1.07e-14      1.000      -0.141       0.141
np.log(expendA) -7.133e-15      0.392  -1.82e-14      1.000      -0.774       0.774
democA            1.11e-15      1.407   7.89e-16      1.000      -2.777       2.777
np.log(expendB)   1.61e-15      0.397   4.05e-15      1.000      -0.785       0.785
==============================================================================
Omnibus:                        6.304   Durbin-Watson:                   1.525
Prob(Omnibus):                  0.043   Jarque-Bera (JB):                6.030
Skew:                           0.448   Prob(JB):                       0.0491
Kurtosis:                       3.182   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R^2=SSE/SST=Σ(ŷ-ȳ)^2/SST.Since the sum of all residuals equal to 0, ȳ=0.For ŷ, ŷ=y-residual(y), since y is the residual from last model which does not explain the variation of last model at all, the residual of y is the same as y which does not explain the model at all. Then ŷ=0, so SSE=0, so R^2=0.

### 2/b)

In [13]:
mydata1["res2"] = Residual**2

aux_reg = smf.ols('res2~prtystrA+np.log(expendA)+democA+np.log(expendB)', mydata1).fit()

f = aux_reg.fvalue
fp = aux_reg.f_pvalue

print("The F-Statistic for the Auxiliary Regression is: "+ str(f) +" and the P-Value is: "+ str(fp))

The F-Statistic for the Auxiliary Regression is: 2.3301126837162744 and the P-Value is: 0.058057514088552704


Since the p-value is greater than 0.05, we can conclude that heteroscedasticity is not present in the model.

### 2/c)

In [14]:
import statsmodels.api as sm
y, X = pt.dmatrices('voteA~prtystrA+np.log(expendA)+democA+np.log(expendB)', mydata1,
                   return_type = 'dataframe')
sm.stats.diagnostic.het_white(Residual, X)

(31.101516025646994,
 0.0032582616087887505,
 2.6807577824828113,
 0.0019733250042551927)

Now, the p-value is 0.002 ,which is less than 0.05, so we reject the null that all of the betas from the secondary
regression are zero, and there is heteroscedasticity in the model from the white test. The evidence of heteroscedasticity is quite strong now since it is far below 0.05.

### 2/d)

In [15]:
reg5 = smf.ols('voteA~prtystrA+np.log(expendA)+democA+np.log(expendB)+shareA', mydata1)
reg5.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.864
Method:                 Least Squares   F-statistic:                     219.1
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           2.09e-71
Time:                        18:09:39   Log-Likelihood:                -557.94
No. Observations:                 173   AIC:                             1128.
Df Residuals:                     167   BIC:                             1147.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          14.4431      4.629      3.120      0.002       5.304      23.582
prtystrA            0.2214      0.058      3.790      0.000       0.106       0.337
np.log(expendA)    -0.6853      0.775     -0.885      0.378      -2.215       0.844
democA              2.3664      1.161      2.038      0.043       0.074       4.659
np.log(expendB)     0.5737      0.811      0.707      0.480      -1.028       2.175
shareA              0.4765      0.052      9.167      0.000       0.374       0.579
==============================================================================
Omnibus:                       28.189   Durbin-Watson:                   1.732
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               76.764
Skew:                           0.637   Prob(JB):                     2.14e-17
Kurtosis:                       6.004   Cond. No.                         767.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [16]:
reg6= smf.ols('voteA~prtystrA+np.log(expendA)+democA+np.log(expendB)+shareA', mydata1)
reg6.fit(cov_type = 'HC0').summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.868
Model:                            OLS   Adj. R-squared:                  0.864
Method:                 Least Squares   F-statistic:                     385.2
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           1.07e-89
Time:                        18:09:39   Log-Likelihood:                -557.94
No. Observations:                 173   AIC:                             1128.
Df Residuals:                     167   BIC:                             1147.
Df Model:                           5                                         
Covariance Type:                  HC0                                         
===================================================================================
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          14.4431      4.865      2.969      0.003       4.908      23.978
prtystrA            0.2214      0.052      4.231      0.000       0.119       0.324
np.log(expendA)    -0.6853      0.521     -1.315      0.189      -1.707       0.336
democA              2.3664      1.166      2.030      0.042       0.082       4.651
np.log(expendB)     0.5737      0.672      0.854      0.393      -0.743       1.890
shareA              0.4765      0.044     10.901      0.000       0.391       0.562
==============================================================================
Omnibus:                       28.189   Durbin-Watson:                   1.732
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               76.764
Skew:                           0.637   Prob(JB):                     2.14e-17
Kurtosis:                       6.004   Cond. No.                         767.
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC0)
"""

The coefficients, standard errors,and significances do not change much after using the robust standard errors procedure.

In [17]:
reg7 = smf.ols('voteA~prtystrA+np.log(expendA)+democA+np.log(expendB)+shareA+district', mydata1)
reg7.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.872
Model:                            OLS   Adj. R-squared:                  0.867
Method:                 Least Squares   F-statistic:                     188.4
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           2.22e-71
Time:                        18:09:39   Log-Likelihood:                -555.13
No. Observations:                 173   AIC:                             1124.
Df Residuals:                     166   BIC:                             1146.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          14.6933      4.569      3.216      0.002       5.672      23.714
prtystrA            0.2323      0.058      4.016      0.000       0.118       0.346
np.log(expendA)    -0.8054      0.766     -1.051      0.295      -2.318       0.707
democA              2.6464      1.152      2.297      0.023       0.372       4.921
np.log(expendB)     0.6505      0.801      0.812      0.418      -0.931       2.232
shareA              0.4842      0.051      9.420      0.000       0.383       0.586
district           -0.1263      0.054     -2.343      0.020      -0.233      -0.020
==============================================================================
Omnibus:                       27.429   Durbin-Watson:                   1.790
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               74.224
Skew:                           0.620   Prob(JB):                     7.63e-17
Kurtosis:                       5.960   Cond. No.                         772.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [18]:
mydata1["ehatsq"] = reg7.fit().resid**2

# estimate weights
w_est = smf.ols('np.log(ehatsq) ~ prtystrA+np.log(expendA)+democA+np.log(expendB)+shareA+district', data = mydata1).fit()

vari = np.exp(w_est.fittedvalues) #estimated variances
w = 1/vari**2

fgls =smf.wls('voteA~ prtystrA+np.log(expendA)+democA+np.log(expendB)+shareA+district', mydata1, weights = w).fit()

fgls.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            WLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.925
Model:                            WLS   Adj. R-squared:                  0.922
Method:                 Least Squares   F-statistic:                     341.8
Date:                Mon, 14 Nov 2022   Prob (F-statistic):           1.13e-90
Time:                        18:09:39   Log-Likelihood:                -565.58
No. Observations:                 173   AIC:                             1145.
Df Residuals:                     166   BIC:                             1167.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          14.9168      3.644      4.094      0.000       7.723      22.111
prtystrA            0.2653      0.054      4.912      0.000       0.159       0.372
np.log(expendA)    -0.7503      0.541     -1.386      0.168      -1.819       0.318
democA              3.1136      0.988      3.152      0.002       1.164       5.064
np.log(expendB)     0.3598      0.510      0.705      0.482      -0.648       1.368
shareA              0.4680      0.039     11.886      0.000       0.390       0.546
district           -0.1593      0.050     -3.178      0.002      -0.258      -0.060
==============================================================================
Omnibus:                       68.924   Durbin-Watson:                   1.957
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              440.050
Skew:                           1.302   Prob(JB):                     2.78e-96
Kurtosis:                      10.367   Cond. No.                         832.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The coefficients, standard errors, and p-values do not change a lot after using the FGLS.